### PR TITLE
Hive: turn off the stats gathering when iceberg.hive.keep.stats is false

### PR DESCRIPTION
**Pull request by stargrey102**
Hive engine collects the stats by traversing the folder to count number of files and size when the table is created and hive.stats.autogather is turned on. The operation can be expensive for a large table while for iceberg table the stats is not needed to be stored on HMS. When the iceberg.hive.keep.stats is set to false, add table parameter "DO_NOT_UPDATE_STATS" so that Hive engine won't collect the stats.

### CODE REVIEW

**Comment by pvary** [comment]
@deniskuzZ: what would be the effect of this change to the Hive integration?

**Comment by stargrey102** [comment]
Some hive tested failed due to a new table property is added: https://github.com/apache/iceberg/pull/10148/commits/27d3c8a8f72e08cd65d4a226e7daf362527b5f7c
Would you mind helping me trigger the integration tests please? @pvary

**Comment by deniskuzZ** [comment]
AFAIK, autogater doesn't even work in Hive. After some operations like insert, we issue an extra stats update task that persists column stats changes either to the HMS or puffin file ("hive.iceberg.stats.source", "iceberg")

**Comment by stargrey102** [comment]
Hi @deniskuzZ:  for more info: we identified the autogather executed (controlled by hive.stats.autogather) to collect stats from Hive when committing a new Iceberg table to Hive (Hive version 3.1.3: https://github.com/apache/hive/blob/rel/release-3.1.3/standalone-metastore/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStore.java#L1868) and that is the motivation of this PR. 
There are two configs:
`hive.stats.autogather` and `hive.stats.column.autogather`
and this pr only tries to disable the first one but not cols stats.

**Comment by deniskuzZ** [comment]
@stargrey102, have you checked the same in Hive-4.0? 
see HIVE-27355, https://github.com/apache/hive/pull/4348

**Comment by stargrey102** [comment]
@deniskuzZ  thank you for the link. HiveOperationsBase uses HiveMetastore client when creating the Iceberg table: https://github.com/apache/iceberg/blob/main/hive-metastore/src/main/java/org/apache/iceberg/hive/HiveOperationsBase.java#L75 while the Hive fix pointed to https://github.com/difin/hive/blob/f96b586c1f338d13b91049a54da09018b3b84723/ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java
Another thought would be if we are not able to upgrade to Hive 4, we would still need to patch it from Iceberg.

**Comment by pvary** [comment]
Thanks @deniskuzZ for the info! Good to know that it will not hurt the Hive integration, to have this PR in.

@stargrey102: Could we create a test which actually checks that the stats are not collected?

File: mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergWithHiveAutogatherEnable.java [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+  private TestTables.TestTableType testTableType;
+
+  @TempDir private Path temp;
+
+  @BeforeAll
+  public static void beforeClass() {
+    // The hive configuration HIVESTATSAUTOGATHER must be set to true from hive engine
+    shell =
+        HiveIcebergStorageHandlerTestUtils.shell(
+            ImmutableMap.of(HiveConf.ConfVars.HIVESTATSAUTOGATHER.varname, "true"));
```
Comment (stargrey102) [inline mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergWithHiveAutogatherEnable.java:75]: The Hive shell sets the hive stats autogather to false as default: https://github.com/apache/iceberg/blob/d047192e01f67b317228f3e075a6ef796f64c180/mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveShell.java#L194
and here to set to true to that we can test this change

**Comment by stargrey102** [comment]
Hi @pvary, sure. I added a test with 2 cases: keep or not keep hive stats.
Since the hive engine set hivestatsautogather to false by default: https://github.com/apache/iceberg/blob/d047192e01f67b317228f3e075a6ef796f64c180/mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveShell.java#L194
I created a new test class instead of adding the test to existing classes, to avoid to impact any existing tests.
Let me know if there is better way to do it. Thanks a lot!

File: mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergWithHiveAutogatherEnable.java [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+  @AfterAll
+  public static void afterClass() throws Exception {
+    shell.stop();
+  }
+
+  @BeforeEach
+  public void before() throws IOException {
+    testTables = HiveIcebergStorageHandlerTestUtils.testTables(shell, testTableType, temp);
+    HiveIcebergStorageHandlerTestUtils.init(shell, testTables, temp, "mz");
```
Comment (pvary) [inline mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergWithHiveAutogatherEnable.java:86]: nit: "mz"? Maybe "mr" 😄

File: mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergWithHiveAutogatherEnable.java [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+    TestTables hiveStatsDisabledTestTables =
+        HiveIcebergStorageHandlerTestUtils.testTables(shell, testTableType, temp);
+
+    TableIdentifier identifierWithoutStats =
+        TableIdentifier.of("default", "customers_without_stats");
+
+    // To validate the stats augother is disabled from Hive engine, the creation of iceberg table
+    // cannot have any
+    // records. Otherwise, the table parameters TOTAL_SIZE and NUM_FILES are added by Iceberg when
+    // inserting records.
```
Comment (pvary) [inline mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergWithHiveAutogatherEnable.java:107]: nit: Could we format this a bit better?

File: mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergWithHiveAutogatherEnable.java [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+    shell.metastore().hiveConf().set(ConfigProperties.KEEP_HIVE_STATS, StatsSetupConst.TRUE);
+    TestTables keepHiveStatsTestTables =
+        HiveIcebergStorageHandlerTestUtils.testTables(shell, testTableType, temp);
+
+    TableIdentifier identifierWithStats = TableIdentifier.of("default", "customers_with_stats");
+
+    // To validate the stats augother is enabled from Hive engine, the creation of iceberg table
+    // cannot have any
+    // records. Otherwise, the table parameters TOTAL_SIZE and NUM_FILES are added by Iceberg when
+    // inserting records.
```
Comment (pvary) [inline mr/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergWithHiveAutogatherEnable.java:152]: nit: Formatting

**Review by pvary — APPROVED** [review]
Small formatting comments, but looks good to me

**Comment by stargrey102** [comment]
Hi @pvary  thank you for the review, typo and format have been fixed. Do you think if it can be merged?

**Comment by pvary** [comment]
Thanks for @stargrey102 for the PR and @deniskuzZ for the help during the review!

**Pull request merged**